File: lua/ProfilesApi.lua
local MakeRequest = require(script.Parent.HTTP).MakePlayFabApiCall
local Promise = require(script.Parent.Promise)
local Settings = require(script.Parent.Settings)
local TS = require(script.Parent.TS)

local GetGlobalPolicy = Promise.promisify(function(request)
	if Settings._internalSettings.entityToken == "" then
		error("Must have Settings._internalSettings.entityToken set to call this method")
	end

	return TS.await(MakeRequest("/Profile/GetGlobalPolicy", request, "X-EntityToken", Settings._internalSettings.entityToken))
end)

local GetProfile = Promise.promisify(function(request)
	if Settings._internalSettings.entityToken == "" then
		error("Must have Settings._internalSettings.entityToken set to call this method")
	end

	return TS.await(MakeRequest("/Profile/GetProfile", request, "X-EntityToken", Settings._internalSettings.entityToken))
end)

local GetProfiles = Promise.promisify(function(request)
	if Settings._internalSettings.entityToken == "" then
		error("Must have Settings._internalSettings.entityToken set to call this method")
	end

	return TS.await(MakeRequest("/Profile/GetProfiles", request, "X-EntityToken", Settings._internalSettings.entityToken))
end)

local GetTitlePlayersFromMasterPlayerAccountIds = Promise.promisify(function(request)
	if Settings._internalSettings.entityToken == "" then
		error("Must have Settings._internalSettings.entityToken set to call this method")
	end

	return TS.await(MakeRequest("/Profile/GetTitlePlayersFromMasterPlayerAccountIds", request, "X-EntityToken", Settings._internalSettings.entityToken))
end)

local SetGlobalPolicy = Promise.promisify(function(request)
	if Settings._internalSettings.entityToken == "" then
		error("Must have Settings._internalSettings.entityToken set to call this method")
	end

	return TS.await(MakeRequest("/Profile/SetGlobalPolicy", request, "X-EntityToken", Settings._internalSettings.entityToken))
end)

local SetProfileLanguage = Promise.promisify(function(request)
	if Settings._internalSettings.entityToken == "" then
		error("Must have Settings._internalSettings.entityToken set to call this method")
	end

	return TS.await(MakeRequest("/Profile/SetProfileLanguage", request, "X-EntityToken", Settings._internalSettings.entityToken))
end)

local SetProfilePolicy = Promise.promisify(function(request)
	if Settings._internalSettings.entityToken == "" then
		error("Must have Settings._internalSettings.entityToken set to call this method")
	end

	return TS.await(MakeRequest("/Profile/SetProfilePolicy", request, "X-EntityToken", Settings._internalSettings.entityToken))
end)

return {
	GetGlobalPolicy = GetGlobalPolicy,
	GetProfile = GetProfile,
	GetProfiles = GetProfiles,
	GetTitlePlayersFromMasterPlayerAccountIds = GetTitlePlayersFromMasterPlayerAccountIds,
	SetGlobalPolicy = SetGlobalPolicy,
	SetProfileLanguage = SetProfileLanguage,
	SetProfilePolicy = SetProfilePolicy,
}

File: src/ProfilesApi.ts
/// <reference path="ProfilesApiTypes.d.ts" />

import MakeRequest from './HTTP'
import Settings from './Settings'

/**
 * Gets the global title access policy
 * https://docs.microsoft.com/rest/api/playfab/profiles/account-management/getglobalpolicy
 */
export const GetGlobalPolicy = async (request: GetGlobalPolicyRequest) => {
    if (Settings._internalSettings.entityToken === '') throw 'Must have Settings._internalSettings.entityToken set to call this method'
    const result = await MakeRequest<GetGlobalPolicyResponse>('/Profile/GetGlobalPolicy', request, 'X-EntityToken', Settings._internalSettings.entityToken)
    return result
}

/**
 * Retrieves the entity's profile.
 * https://docs.microsoft.com/rest/api/playfab/profiles/account-management/getprofile
 */
export const GetProfile = async (request: GetEntityProfileRequest) => {
    if (Settings._internalSettings.entityToken === '') throw 'Must have Settings._internalSettings.entityToken set to call this method'
    const result = await MakeRequest<GetEntityProfileResponse>('/Profile/GetProfile', request, 'X-EntityToken', Settings._internalSettings.entityToken)
    return result
}

/**
 * Retrieves the entity's profile.
 * https://docs.microsoft.com/rest/api/playfab/profiles/account-management/getprofiles
 */
export const GetProfiles = async (request: GetEntityProfileRequest) => {
    if (Settings._internalSettings.entityToken === '') throw 'Must have Settings._internalSettings.entityToken set to call this method'
    const result = await MakeRequest<GetEntityProfileResponse>('/Profile/GetProfiles', request, 'X-EntityToken', Settings._internalSettings.entityToken)
    return result
}

/**
 * Retrieves the title player accounts associated with the given master player account.
 * https://docs.microsoft.com/rest/api/playfab/profiles/account-management/gettitleplayersfrommasterplayeraccountids
 */
export const GetTitlePlayersFromMasterPlayerAccountIds = async (request: GetTitlePlayersFromMasterPlayerAccountIdsRequest) => {
    if (Settings._internalSettings.entityToken === '') throw 'Must have Settings._internalSettings.entityToken set to call this method'
    const result = await MakeRequest<GetTitlePlayersFromMasterPlayerAccountIdsResponse>('/Profile/GetTitlePlayersFromMasterPlayerAccountIds', request, 'X-EntityToken', Settings._internalSettings.entityToken)
    return result
}

/**
 * Sets the global title access policy
 * https://docs.microsoft.com/rest/api/playfab/profiles/account-management/setglobalpolicy
 */
export const SetGlobalPolicy = async (request: SetGlobalPolicyRequest) => {
    if (Settings._internalSettings.entityToken === '') throw 'Must have Settings._internalSettings.entityToken set to call this method'
    const result = await MakeRequest<SetGlobalPolicyResponse>('/Profile/SetGlobalPolicy', request, 'X-EntityToken', Settings._internalSettings.entityToken)
    return result
}

/**
 * Updates the entity's language. The precedence hierarchy for communication to the player is Title Player Account
 * language, Master Player Account language, and then title default language if the first two aren't set or supported.
 * https://docs.microsoft.com/rest/api/playfab/profiles/account-management/setprofilelanguage
 */
export const SetProfileLanguage = async (request: SetProfileLanguageRequest) => {
    if (Settings._internalSettings.entityToken === '') throw 'Must have Settings._internalSettings.entityToken set to call this method'
    const result = await MakeRequest<SetProfileLanguageResponse>('/Profile/SetProfileLanguage', request, 'X-EntityToken', Settings._internalSettings.entityToken)
    return result
}

/**
 * Sets the profiles access policy
 * https://docs.microsoft.com/rest/api/playfab/profiles/account-management/setprofilepolicy
 */
export const SetProfilePolicy = async (request: SetEntityProfilePolicyRequest) => {
    if (Settings._internalSettings.entityToken === '') throw 'Must have Settings._internalSettings.entityToken set to call this method'
    const result = await MakeRequest<SetEntityProfilePolicyResponse>('/Profile/SetProfilePolicy', request, 'X-EntityToken', Settings._internalSettings.entityToken)
    return result
}